docs/NetEbpfExtMultiAttach.md

```diff
@@ -0,0 +1,126 @@
+
+
+## Synchronization
+In the new design, access to the list of clients in the filter context and the list of filter contexts in the provider
+context needs to be synchronized. The 2 flows that need synchronization specifically are:
+1. Program invocation
+2. Client attach / detach callback.
+
+To provide synchronization, the hook provider will maintain a *dispatch* level RW lock to synchronize access to these
+lists. (A dispatch level lock is chosen here as the WFP callouts can be invoked at either PASSIVE_LEVEL or DISPATCH_LEVEL).
+1. Attach and detach callbacks flow will acquire this lock in exclusive mode.
+2. Program invocation flow acquires this lock in shared mode.
+
+This approach allows multiple program invocations at the same time, as well as protecting the list when a client is
+being attached / detached. Client / attach operations are expected to be much less frequent than program invocations.
+This approach eliminates any need for rundown reference for the client context. Since the list of client contexts for a
+filter context is synchronized via a lock, it is always guaranteed that in the client detach flow, if the dispatch level
+lock is acquired, there is no program invocation in progress for that program. Also, once the client context is removed
+from the list of clients in the filter context, no new classify callback will find this program in the list, hence no
+new invocation can also happen.
+
+Along with the above synchronization, there is also a need to **serialize multiple attach and detach operations callbacks**,
+as the whole flow of creating filter context, adding filter context to the provider list, and configuring WFP filters
+needs to happen atomically (as mentioned in the current design section, there is bug currently where a lock is not held
+for the whole duration of attach / detach callback).
+
+As the above mentioned dispatch level RW already serializes attach / detach callbacks for a hook provider, it could have
+solved this serialization problem also. But there is a problem with using dispatch level locks: WFP APIs for adding
+filters require PASSIVE_LEVEL, hence same DISPATCH_LEVEL lock cannot be used to serialize the attach and detach
+operations. To address this, a separate PASSIVE lock will be maintained in the provider context to serialize attach and
+detach operations. As a result of this, in attach and detach operations, the flow acquires both the DISPATCH_LEVEL lock
+and the PASSIVE_LEVEL lock. In the program invocation flow, only the DISPATCH_LEVEL lock is acquired.
+
+**Expected flow for client attach callback**:
+1. Acquire passive lock --> Serializes multiple attach / detach calls to the same hook provider.
+2. Acquire dispatch lock (exclusive). --> Synchronizes access to the client and filter context lists.
+3. Check if a matching filter context already exists. If it exists, insert the new client context, and return.
+4. If no matching filter context exists, release dispatch lock.
```

Comment:
What happens if the state changes after this lock is released? Won't this permit multiple callers to create the WFP context?

Reply:
The passive lock (acquired in step 1) still serializes attach / detach calls to this provider.